treewide: Start to break up static overlay

Ericson2314 opened the pull request:
###### Motivation for this change

We can use use `stdenv.hostPlatform.isStatic` instead, and move the
logic per package. The least opionated benefit of this is that it makes
it much easier to replace packages with modified ones, as there is no
longer any issue of overlay order.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CC @FRidh @matthewbauer

Ericson2314 committed:
treewide: Start to break up static overlay

We can use use `stdenv.hostPlatform.isStatic` instead, and move the
logic per package. The least opionated benefit of this is that it makes
it much easier to replace packages with modified ones, as there is no
longer any issue of overlay order.

CC @FRidh @matthewbauer

Ericson2314 committed:
gcc 8, 10: tabs -> spaces

FRidh commented:
@matthewbauer suggested

> I'm not sure if I exactly like "isStatic". Probably a better name is isStaticLibraries and isStaticBinaries for each of the stdenv adapters.

Any comment on that?

nh2 commented:
+1 on having a standard way of querying whether a package should build static/shared libaries and static/dynamic executables, but:

> @matthewbauer suggested
> 
> > I'm not sure if I exactly like "isStatic". Probably a better name is isStaticLibraries and isStaticBinaries for each of the stdenv adapters.

I heavily agree with this.

I think `isStatic` is not enough, and not a good name. For my use case (building static Haskell executables), I (and some others, see #61575) need packages to build:

* static libraries (`.a` files)
* dynamic libraries as well (`.so` files) so that TemplateHaskell can load them
* I do not need them to build static executables (e.g. I don't need a statically linked `curl` exe; that just makes it more complicated)

Others want only `.a` (`.so` being absent), and again others want static executables.

**These multiple choices cannot be expressed with a single `isStatic` bool.**

Consequently, I think that the

> `, enableShared ? !stdenv.hostPlatform.isStatic`

proposed in this PR, is not good, and that **we should rather have names like `buildStaticLibs`, `buildSharedLibs`** (which can be used in any combination, except perhaps `false`/`false` because then we would not build anything), and `buildStaticExes`.

Ericson2314 commented:
@nh2 

> I do not need them to build static executables (e.g. I don't need a statically linked curl exe; that just makes it more complicated)

remember the `hostPlatform` part. For `pkgsCross`, `stdenv.buildPlatform.isStatic = false` while `stdenv.hostPlatform.isStatic = true`. A build-time `curl` won't be statically linked.

But stepping back a bit, `stdenv.buildPlatform.isStatic` already exists, and the overlay has all the same issues you describe as this PR. I would think actually think this is is closer to what you want, because it's much easier to make more parameters than make more overlays.

matthewbauer commented on pkgs/development/compilers/gcc/10/default.nix:
```diff
@@ -7,7 +7,7 @@
 , langJit ? false
 , staticCompiler ? false
-, enableShared ? true
+, enableShared ? !stdenv.hostPlatform.isStatic
```
Don't do this!!! We want the toolchain to be able to produce both static & shared. Otherwise we get mass rebuilds for no reason. GCC & LLVM should be capable of producing both shared & static binaries as we need them.

Ericson2314 replied:
OK not doing this for now.

matthewbauer commented on pkgs/os-specific/darwin/apple-source-releases/libiconv/default.nix:
```diff
@@ -1,6 +1,6 @@
 { stdenv, appleDerivation, lib
-, enableStatic ? stdenv.targetPlatform.isiOS
-, enableShared ? !stdenv.targetPlatform.isiOS
+, enableStatic ? stdenv.targetPlatform.isStatic
```
This should be hostPlatform

matthewbauer commented:
So there's definitely a benefit to putting everything in one place - most maintainers don't know about static overlay so having lots of isStatic conditionals in their definitions will make them aware they need to test this. It also means we expect pkgsStatic to really be supported everywhere and it's more than just an after thought. I think we'll need to make sure Nixpkgs maintainers for these packages are okay with that extra burden.

@Ericson2314 Would you consider moving "isStatic" to "stdenv" instead of "stdenv.hostPlatform"? Conceptually, I think this is an important distinction because:

- stdenv.hostPlatform describes what ABI and toolchain we are using. You can never link packages from two different hostPlatform's together. But "static" / "shared" is much more flexible. You can always have some packages that are static and others that aren't but they link together fine. This hasn't been explored much, but I think it's important to leave this open especially with things that need shared linking like GHC / Template Haskell.
- "crossSystem" should be limited to what's needed to make a cross toolchain (arch, vendor, kernel, environment, linker, compilers, etc.). That's the part of the stdenv that's special for cross-compiling. All other static logic can be shared with native compilation.

Ericson2314 commented:
@matthewbauer I don't think I really believe in a "platform vs rest" of config distinction.

> You can never link packages from two different hostPlatform's together.

But you can:

- Arm hard float with soft-float ABI exists to be compatible
- i486 -> i586 -> i686, etc.
- RISC-V's extension modes
- WASM presumably has feature flags for various draft things
- Fat binaries / bootloaders / runtime `cpuid()` type thing

True, nothing in this world has discrete interfaces. There is always a comparability relation.

> should be limited to what's needed to make a cross toolchain (arch, vendor, kernel, environment, linker, compilers, etc.).

I don't think there is a firm line where "toolchain" ends. And `isStatic` could very well influence the toolchain itself by deciding whether to build a dynamic linker.

> "stdenv" is where the stdenv adapter is applied, so it's a natural point to determine what our stdenv.mkDerivation is going to do. This would more clearly separate the stdenv toolchain from the crossSystem description.

Well, I also am not sure we we want to stick with a stdenv adapter long-term either. We could use `isStatic` with the default autoconf / cmake / meson flags just like we do with the rest of `*Platform`

-------

Finally, it's totally reasonable one would look at `stdenv.buildPlatform.isStatic` or `stdenv.targetPlatform.isStatic`, and that's a bit more cumbersome with `stdenv.isStatic`, especially because `targetPackages.stdenv.isStatic` is more likely to hit infinite recursion.

Ericson2314 commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -2,7 +2,7 @@
 # Enabling all targets increases output size to a multiple.
 , withAllTargets ? false, libbfd, libopcodes
-, enableShared ? true
+, enableShared ? !stdenv.hostPlatform.isStatic
```
This one is really about the libs ld itself uses, not target platform stuff --- `hostPlatform` from my sed is actually correct --- so I think it's OK?

Ericson2314 committed:
gcc: Always pass `--enable-shared` by default

I am actually a bit skeptical about this, but @matthewbauer makes the
case for this in
https://github.com/NixOS/nixpkgs/pull/107238#discussion_r546454453 and
I'm happy to go with it not being as in the loop on static linking stuff
as he is.

Ericson2314 commented:
OK if no one ways anything, I think i want to merge it because this rots decently.

 - I did Matt's change for GCC, though I have some misgivings.
 - The `isStatic` changes are technically orthogonal, so we can always do them after.

FRidh commented:
Could you rebase on master instead of having a merge.

Ericson2314 commented on pkgs/development/libraries/gmp/5.1.x.nix:
```diff
@@ -1,4 +1,7 @@
-{ stdenv, fetchurl, m4, cxx ? true, withStatic ? true }:
```
I think this should have been `withStatic ? false` all along, without a comment to justify it.
